### ISSUE TITLE
Add an elegant project switcher like Sublime's

### ISSUE THREAD

**Issue by brismuth** [thomaswilburn/Caret#456]
It would be awesome if there was a quick project switcher like Sublime's or [Zed](https://github.com/zedapp/zed)'s. 

Since projects are likely not stored in the same locations across different machines, the list of available projects would probably have to be kept locally on disk. One way of doing this would be to store all of the project settings that are currently in different files in a single shared file. Then users wouldn't have to worry about picking a place to save project files; they could just open a new folder and it could become a new project. Zed handles it that way and I think it is incredibly convenient.

**Comment by thomaswilburn** [comment]
The thing to keep in mind with this--and don't get me wrong, it sounds really great--is that we can't always actually "re-open" a file or tab if Chrome decides (for a variety of reasons) that it is revoking our access (see [retainEntry's documentation](https://developer.chrome.com/apps/fileSystem#method-retainEntry)). We're only allowed to keep 500 files retained, although I don't believe that directories count against that. Caret automatically retains whatever files it opens (in case it gets shut down without warning), I don't know that we can always predict that we'll hit the limit. For that reason, I've been wary about adding too many features like a "recent files" menu, because I don't like adding features that won't work reliably. I think we would definitely need to test a feature like this carefully before rolling it out.

**Comment by brismuth** [comment]
Caret already supports opening a previously existing project file. It seems to me that the only thing the project switcher would change would be that the UI for opening projects would be much easier to use. Is that an incorrect understanding?

**Comment by thomaswilburn** [comment]
When we re-open the file, it contains the file retention IDs for any of the directories that were in the project, and Caret tries to restore them. But this isn't guaranteed--actually, restoring directories in a project fails all the time, which is really annoying, just because for some reason Chrome has expired our retained file privileges for that ID.

Adding a list of projects means we have two layers now: one layer for the project files, and one layer for the files referenced by those projects. That makes me nervous. Unless you're proposing that we just move all project files into a single file somewhere on the hard drive? I may have misunderstood your proposal, to be honest.

Either way, I'd love to be wrong about this, so don't let me discourage you. But I did want to bring up some of the reasons why it is built in the way that it is.

**Comment by brismuth** [comment]
Right, I was suggesting storing all project files/references in a single local file. Do you have any kind of logging or stats on how often it fails to open directories? It is a pity that it happens at all.

**Comment by thomaswilburn** [comment]
No, no stats, unfortunately. I just remember that when I was working heavily on the initial implementation for projects, I would see it happen enough for it to be annoying, but not often enough that it was easy to reproduce. Like I said, maybe it's not a big deal, and maybe it works better now than it used to (I don't think the retainFile API was very old at the time).

**Comment by brismuth** [comment]
Do you know if it's possible to collect usage data and error logs, or would that violate the terms of service of the Chrome Web Store? I feel like it would be useful to have that kind of data to decide whether or not projects like this are feasible or worth it.

**Comment by thomaswilburn** [comment]
It's certainly possible, as long as the tracking domain is whitelisted in the manifest. I've resisted doing it so far, because I like that Caret really does work completely offline and doesn't do any user tracking at all, no matter how anonymized it's supposed to be. I never liked that Atom and Zed are firing Google Analytics events constantly while I'm using them.
